rclcpp/src/rclcpp/node_interfaces/node_parameters.cpp

```diff
@@ -185,6 +188,88 @@ __lockless_has_parameter(
   return parameters.find(name) != parameters.end();
 }
 
+// see https://en.cppreference.com/w/cpp/types/numeric_limits/epsilon
+#define ULP 100
+RCLCPP_LOCAL
+bool
+__are_doubles_equal(double x, double y, int ulp = ULP)
+{
+  return std::abs(x - y) <= std::numeric_limits<double>::epsilon() * std::abs(x + y) * ulp;
+}
+
+RCLCPP_LOCAL
+bool
+__check_parameter_value_in_range(
+  const rcl_interfaces::msg::ParameterDescriptor & descriptor,
+  const rclcpp::ParameterValue & value)
+{
+  if (!descriptor.integer_range.empty() && value.get_type() == rclcpp::PARAMETER_INTEGER) {
+    int64_t v = value.get<int64_t>();
+    auto integer_range = descriptor.integer_range.at(0);
+    if ((v < integer_range.from_value) || (v > integer_range.to_value)) {
+      return false;
+    }
+    if (v == integer_range.to_value) {
+      return true;
+    }
+    if (integer_range.step == 0) {
+      return true;
+    }
+    if (((v - integer_range.from_value) % integer_range.step) == 0) {
+      return true;
+    }
+    return false;
+  }
+
+  if (!descriptor.floating_point_range.empty() && value.get_type() == rclcpp::PARAMETER_DOUBLE) {
+    double v = value.get<double>();
+    auto fp_range = descriptor.floating_point_range.at(0);
+    if (__are_doubles_equal(v, fp_range.from_value) ||
+      __are_doubles_equal(v, fp_range.to_value))
+    {
+      return true;
+    }
+    if ((v < fp_range.from_value) || (v > fp_range.to_value)) {
+      return false;
+    }
+    if (fp_range.step == 0.0) {
+      return true;
+    }
+    int truncated_div = (v - fp_range.from_value) / fp_range.step;
```

Comment:
Instead of checking for `truncated_div` and `trincated_div + `, you could `round` the result here instead of truncating.

I'll exaggerate the math a bit to illustrate the point. Imagine `__are_doubles_equal` allows a tolerance of 0.2, that you have a step of 0.5 from 0 to 10, and you set 4.9 as the value. If you truncate, the `truncated_div` will be `floor((4.9 - 0) / 0.5)`, which would be `floor(9.8) = 9`. When you apply `__are_doubles_equal` in the line below, `from + 9 * step` will be 4.5, which is too far away from the value for the check to pass. When you apply `__are_doubles_equal` with `truncated_div + 1`, it will pass.

If instead of truncating by casting to int you round this calculation, 4.9 would round to 10 steps, and the first check in the line below would pass (which I think is a bit simpler than applying `__are_doubles_equal` twice).